File: ios/Sources/YandexAdsPlugin/YandexAdsPlugin.swift
import Capacitor
import Foundation
import UIKit
import YandexMobileAds

@objc(YandexAdsPlugin)
public class YandexAdsPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "YandexAdsPlugin"
    public let jsName = "YandexAds"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setUserConsent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setAgeRestricted", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setLocationTracking", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "showBanner", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removeBanner", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "prepareInterstitial", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isInterstitialReady", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "showInterstitial", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "prepareRewarded", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isRewardedReady", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "showRewarded", returnType: CAPPluginReturnPromise)
    ]

    private let interstitialAdLoader = InterstitialAdLoader()
    private let rewardedAdLoader = RewardedAdLoader()
    private var bannerAdView: BannerAdView?
    private var bannerLoadCall: CAPPluginCall?
    private var bannerAdUnitID = ""
    private var interstitialAd: InterstitialAd?
    private var rewardedAd: RewardedAd?
    private var interstitialAdUnitID = ""
    private var rewardedAdUnitID = ""
    private var initialized = false
    private var initializing = false
    private var interstitialLoading = false
    private var rewardedLoading = false

    @objc func initialize(_ call: CAPPluginCall) {
        runOnMain { self.initializeOnMain(call) }
    }

    @MainActor private func initializeOnMain(_ call: CAPPluginCall) {
        call.getBool("userConsent").map(YandexAds.setUserConsent)
        call.getBool("ageRestricted").map(YandexAds.setAgeRestricted)
        call.getBool("locationTracking").map(YandexAds.setLocationTracking)
        if call.getBool("enableLogging") == true {
            YandexAds.enableLogging()
        }

        if initialized {
            call.resolve(["version": YandexAds.sdkVersion.stringValue])
            return
        }
        guard !initializing else {
            call.reject("Yandex Ads initialization is already in progress", "INITIALIZATION_IN_PROGRESS")
            return
        }

        initializing = true
        YandexAds.initializeSDK { [weak self] in
            guard let self else { return }
            self.initializing = false
            self.initialized = true
            call.resolve(["version": YandexAds.sdkVersion.stringValue])
        }
    }

    @objc func setUserConsent(_ call: CAPPluginCall) {
        runOnMain { self.setBoolean(call, setter: YandexAds.setUserConsent) }
    }

    @objc func setAgeRestricted(_ call: CAPPluginCall) {
        runOnMain { self.setBoolean(call, setter: YandexAds.setAgeRestricted) }
    }

    @objc func setLocationTracking(_ call: CAPPluginCall) {
        runOnMain { self.setBoolean(call, setter: YandexAds.setLocationTracking) }
    }

    @objc func showBanner(_ call: CAPPluginCall) {
        runOnMain { self.showBannerOnMain(call) }
    }

    @MainActor private func showBannerOnMain(_ call: CAPPluginCall) {
        guard requireInitialized(call), let adUnitID = requireAdUnitID(call) else { return }
        guard bannerLoadCall == nil else {
            call.reject("A banner load is already in progress", "LOAD_IN_PROGRESS")
            return
        }
        guard let viewController = bridge?.viewController else {
            call.reject("iOS view controller is unavailable", "VIEW_CONTROLLER_UNAVAILABLE")
            return
        }

        destroyBannerAd()
        bannerAdUnitID = adUnitID
        bannerLoadCall = call
        let width = max(viewController.view.bounds.width, 1)
        let bannerSize = BannerAdSize.sticky(containerWidth: width)
        let banner = BannerAdView(adSize: bannerSize)
        banner.delegate = self
        bannerAdView = banner
        banner.displayAtBottom(in: viewController.view)
        banner.loadAd(with: AdRequest(adUnitID: adUnitID))
    }

    @objc func removeBanner(_ call: CAPPluginCall) {
        runOnMain {
            self.bannerLoadCall?.reject("Banner was removed", "BANNER_REMOVED")
            self.bannerLoadCall = nil
            self.destroyBannerAd()
            call.resolve()
        }
    }

    @objc func prepareInterstitial(_ call: CAPPluginCall) {
        runOnMain { self.prepareInterstitialOnMain(call) }
    }

    @MainActor private func prepareInterstitialOnMain(_ call: CAPPluginCall) {
        guard requireInitialized(call), let adUnitID = requireAdUnitID(call) else { return }
        guard !interstitialLoading else {
            call.reject("An interstitial load is already in progress", "LOAD_IN_PROGRESS")
            return
        }

        destroyInterstitialAd()
        interstitialAdUnitID = adUnitID
        interstitialLoading = true
        interstitialAdLoader.loadAd(with: AdRequest(adUnitID: adUnitID)) { [weak self] result in
            guard let self else { return }
            self.interstitialLoading = false
            switch result {
            case .success(let ad):
                self.interstitialAd = ad
                ad.delegate = self
                self.notifyListeners("interstitialLoaded", data: self.adEvent(adUnitID))
                call.resolve()
            case .failure(let error):
                self.notifyListeners("interstitialFailedToLoad", data: self.errorEvent(adUnitID, error))
                self.reject(call, error: error, code: "INTERSTITIAL_LOAD_FAILED")
            }
        }
    }

    @objc func isInterstitialReady(_ call: CAPPluginCall) {
        runOnMain {
            call.resolve(["ready": self.interstitialAd != nil])
        }
    }

    @objc func showInterstitial(_ call: CAPPluginCall) {
        runOnMain { self.showInterstitialOnMain(call) }
    }

    @MainActor private func showInterstitialOnMain(_ call: CAPPluginCall) {
        guard let ad = interstitialAd else {
            call.reject("Interstitial ad is not ready", "AD_NOT_READY")
            return
        }
        guard let viewController = bridge?.viewController else {
            call.reject("iOS view controller is unavailable", "VIEW_CONTROLLER_UNAVAILABLE")
            return
        }
        ad.show(from: viewController)
        call.resolve()
    }

    @objc func prepareRewarded(_ call: CAPPluginCall) {
        runOnMain { self.prepareRewardedOnMain(call) }
    }

    @MainActor private func prepareRewardedOnMain(_ call: CAPPluginCall) {
        guard requireInitialized(call), let adUnitID = requireAdUnitID(call) else { return }
        guard !rewardedLoading else {
            call.reject("A rewarded ad load is already in progress", "LOAD_IN_PROGRESS")
            return
        }

        destroyRewardedAd()
        rewardedAdUnitID = adUnitID
        rewardedLoading = true
        rewardedAdLoader.loadAd(with: AdRequest(adUnitID: adUnitID)) { [weak self] result in
            guard let self else { return }
            self.rewardedLoading = false
            switch result {
            case .success(let ad):
                self.rewardedAd = ad
                ad.delegate = self
                self.notifyListeners("rewardedLoaded", data: self.adEvent(adUnitID))
                call.resolve()
            case .failure(let error):
                self.notifyListeners("rewardedFailedToLoad", data: self.errorEvent(adUnitID, error))
                self.reject(call, error: error, code: "REWARDED_LOAD_FAILED")
            }
        }
    }

    @objc func isRewardedReady(_ call: CAPPluginCall) {
        runOnMain {
            call.resolve(["ready": self.rewardedAd != nil])
        }
    }

    @objc func showRewarded(_ call: CAPPluginCall) {
        runOnMain { self.showRewardedOnMain(call) }
    }

    @MainActor private func showRewardedOnMain(_ call: CAPPluginCall) {
        guard let ad = rewardedAd else {
            call.reject("Rewarded ad is not ready", "AD_NOT_READY")
            return
        }
        guard let viewController = bridge?.viewController else {
            call.reject("iOS view controller is unavailable", "VIEW_CONTROLLER_UNAVAILABLE")
            return
        }
        ad.show(from: viewController)
        call.resolve()
    }

    @MainActor private func setBoolean(_ call: CAPPluginCall, setter: (Bool) -> Void) {
        guard let value = call.getBool("value") else {
            call.reject("value is required", "INVALID_ARGUMENT")
            return
        }
        setter(value)
        call.resolve()
    }

    @MainActor private func requireInitialized(_ call: CAPPluginCall) -> Bool {
        guard initialized else {
            call.reject("Call initialize() before loading ads", "NOT_INITIALIZED")
            return false
        }
        return true
    }

    @MainActor private func requireAdUnitID(_ call: CAPPluginCall) -> String? {
        guard let value = call.getString("adUnitId")?.trimmingCharacters(in: .whitespacesAndNewlines), !value.isEmpty else {
            call.reject("adUnitId is required", "INVALID_ARGUMENT")
            return nil
        }
        return value
    }

    @MainActor private func adEvent(_ adUnitID: String) -> [String: Any] {
        ["adUnitId": adUnitID]
    }

    @MainActor private func errorEvent(_ adUnitID: String, _ error: Error) -> [String: Any] {
        let nsError = error as NSError
        return ["adUnitId": adUnitID, "code": nsError.code, "message": nsError.localizedDescription]
    }

    @MainActor private func reject(_ call: CAPPluginCall, error: Error, code: String) {
        call.reject((error as NSError).localizedDescription, code, error)
    }

    @MainActor private func destroyInterstitialAd() {
        interstitialAd?.delegate = nil
        interstitialAd = nil
    }

    @MainActor private func destroyBannerAd() {
        bannerAdView?.delegate = nil
        bannerAdView?.removeFromSuperview()
        bannerAdView = nil
    }

    @MainActor private func destroyRewardedAd() {
        rewardedAd?.delegate = nil
        rewardedAd = nil
    }

    private func runOnMain(_ block: @escaping @MainActor () -> Void) {
        DispatchQueue.main.async(execute: block)
    }
}

@MainActor extension YandexAdsPlugin: BannerAdViewDelegate {
    public func bannerAdViewDidLoad(_ bannerAdView: BannerAdView) {
        let size = bannerAdView.adContentSize()
        let result: [String: Any] = [
            "adUnitId": bannerAdUnitID,
            "width": size.width,
            "height": size.height
        ]
        notifyListeners("bannerLoaded", data: result)
        bannerLoadCall?.resolve(result)
        bannerLoadCall = nil
    }

    public func bannerAdViewDidFailLoading(_ bannerAdView: BannerAdView, error: Error) {
        notifyListeners("bannerFailedToLoad", data: errorEvent(bannerAdUnitID, error))
        bannerLoadCall?.reject((error as NSError).localizedDescription, "BANNER_LOAD_FAILED", error)
        bannerLoadCall = nil
        destroyBannerAd()
    }

    public func bannerAdViewDidClick(_ bannerAdView: BannerAdView) {
        notifyListeners("bannerClicked", data: adEvent(bannerAdUnitID))
    }

    public func bannerAdView(_ bannerAdView: BannerAdView, didTrackImpression impressionData: ImpressionData?) {
        notifyListeners("bannerImpression", data: adEvent(bannerAdUnitID))
    }
}

@MainActor extension YandexAdsPlugin: InterstitialAdDelegate {
    public func interstitialAdDidShow(_ interstitialAd: InterstitialAd) {
        notifyListeners("interstitialShown", data: adEvent(interstitialAdUnitID))
    }

    public func interstitialAdDidDismiss(_ interstitialAd: InterstitialAd) {
        notifyListeners("interstitialDismissed", data: adEvent(interstitialAdUnitID))
        destroyInterstitialAd()
    }

    public func interstitialAdDidClick(_ interstitialAd: InterstitialAd) {
        notifyListeners("interstitialClicked", data: adEvent(interstitialAdUnitID))
    }

    public func interstitialAd(_ interstitialAd: InterstitialAd, didTrackImpression impressionData: ImpressionData?) {
        notifyListeners("interstitialImpression", data: adEvent(interstitialAdUnitID))
    }

    public func interstitialAd(_ interstitialAd: InterstitialAd, didFailToShow error: Error) {
        notifyListeners("interstitialFailedToShow", data: errorEvent(interstitialAdUnitID, error))
        destroyInterstitialAd()
    }
}

@MainActor extension YandexAdsPlugin: RewardedAdDelegate {
    public func rewardedAdDidShow(_ rewardedAd: RewardedAd) {
        notifyListeners("rewardedShown", data: adEvent(rewardedAdUnitID))
    }

    public func rewardedAdDidDismiss(_ rewardedAd: RewardedAd) {
        notifyListeners("rewardedDismissed", data: adEvent(rewardedAdUnitID))
        destroyRewardedAd()
    }

    public func rewardedAdDidClick(_ rewardedAd: RewardedAd) {
        notifyListeners("rewardedClicked", data: adEvent(rewardedAdUnitID))
    }

    public func rewardedAd(_ rewardedAd: RewardedAd, didTrackImpression impressionData: ImpressionData?) {
        notifyListeners("rewardedImpression", data: adEvent(rewardedAdUnitID))
    }

    public func rewardedAd(_ rewardedAd: RewardedAd, didReward reward: Reward) {
        notifyListeners("rewarded", data: [
            "adUnitId": rewardedAdUnitID,
            "amount": reward.amount,
            "type": reward.type
        ])
    }

    public func rewardedAd(_ rewardedAd: RewardedAd, didFailToShow error: Error) {
        notifyListeners("rewardedFailedToShow", data: errorEvent(rewardedAdUnitID, error))
        destroyRewardedAd()
    }
}

File: android/src/main/kotlin/com/barsukstudio/plugins/yandexads/YandexAdsPlugin.kt
package com.barsukstudio.plugins.yandexads

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.yandex.mobile.ads.common.AdError
import com.yandex.mobile.ads.common.AdRequest
import com.yandex.mobile.ads.common.AdRequestError
import com.yandex.mobile.ads.common.ImpressionData
import com.yandex.mobile.ads.common.YandexAds
import com.yandex.mobile.ads.interstitial.InterstitialAd
import com.yandex.mobile.ads.interstitial.InterstitialAdEventListener
import com.yandex.mobile.ads.interstitial.InterstitialAdLoadListener
import com.yandex.mobile.ads.interstitial.InterstitialAdLoader
import com.yandex.mobile.ads.rewarded.Reward
import com.yandex.mobile.ads.rewarded.RewardedAd
import com.yandex.mobile.ads.rewarded.RewardedAdEventListener
import com.yandex.mobile.ads.rewarded.RewardedAdLoadListener
import com.yandex.mobile.ads.rewarded.RewardedAdLoader

@CapacitorPlugin(name = "YandexAds")
class YandexAdsPlugin : Plugin() {
    private var initialized = false
    private var initializingCall: PluginCall? = null

    private var interstitialAdLoader: InterstitialAdLoader? = null
    private var interstitialAd: InterstitialAd? = null
    private var interstitialAdUnitId = ""
    private var interstitialLoadCall: PluginCall? = null

    private var rewardedAdLoader: RewardedAdLoader? = null
    private var rewardedAd: RewardedAd? = null
    private var rewardedAdUnitId = ""
    private var rewardedLoadCall: PluginCall? = null

    @PluginMethod
    fun initialize(call: PluginCall) = runOnMainThread {
        call.getBoolean("userConsent")?.let(YandexAds::setUserConsent)
        call.getBoolean("ageRestricted")?.let(YandexAds::setAgeRestricted)
        call.getBoolean("locationTracking")?.let(YandexAds::setLocationTracking)
        call.getBoolean("enableLogging")?.let(YandexAds::enableLogging)

        if (initialized) {
            call.resolve(versionResult())
            return@runOnMainThread
        }
        if (initializingCall != null) {
            call.reject("Yandex Ads initialization is already in progress", "INITIALIZATION_IN_PROGRESS")
            return@runOnMainThread
        }

        initializingCall = call
        YandexAds.initialize(context) {
            initialized = true
            initializingCall?.resolve(versionResult())
            initializingCall = null
        }
    }

    @PluginMethod
    fun setUserConsent(call: PluginCall) = setBoolean(call, YandexAds::setUserConsent)

    @PluginMethod
    fun setAgeRestricted(call: PluginCall) = setBoolean(call, YandexAds::setAgeRestricted)

    @PluginMethod
    fun setLocationTracking(call: PluginCall) = setBoolean(call, YandexAds::setLocationTracking)

    @PluginMethod
    fun prepareInterstitial(call: PluginCall) = runOnMainThread {
        val adUnitId = requireAdUnitId(call) ?: return@runOnMainThread
        if (!requireInitialized(call)) return@runOnMainThread
        if (interstitialLoadCall != null) {
            call.reject("An interstitial load is already in progress", "LOAD_IN_PROGRESS")
            return@runOnMainThread
        }

        destroyInterstitialAd()
        interstitialAdUnitId = adUnitId
        interstitialLoadCall = call
        if (interstitialAdLoader == null) interstitialAdLoader = InterstitialAdLoader(context)
        val request = AdRequest.Builder(adUnitId).build()
        interstitialAdLoader?.loadAd(request, object : InterstitialAdLoadListener {
            override fun onAdLoaded(interstitialAd: InterstitialAd) {
                this@YandexAdsPlugin.interstitialAd = interstitialAd
                notifyListeners("interstitialLoaded", adEvent(interstitialAdUnitId))
                interstitialLoadCall?.resolve()
                interstitialLoadCall = null
            }

            override fun onAdFailedToLoad(error: AdRequestError) {
                val event = errorEvent(interstitialAdUnitId, error.code, error.description)
                notifyListeners("interstitialFailedToLoad", event)
                interstitialLoadCall?.reject(error.description, "INTERSTITIAL_LOAD_FAILED")
                interstitialLoadCall = null
            }
        })
    }

    @PluginMethod
    fun isInterstitialReady(call: PluginCall) = runOnMainThread {
        call.resolve(JSObject().put("ready", interstitialAd != null))
    }

    @PluginMethod
    fun showInterstitial(call: PluginCall) = runOnMainThread {
        val ad = interstitialAd
        val currentActivity = activity
        if (ad == null) {
            call.reject("Interstitial ad is not ready", "AD_NOT_READY")
            return@runOnMainThread
        }
        if (currentActivity == null) {
            call.reject("Android Activity is unavailable", "ACTIVITY_UNAVAILABLE")
            return@runOnMainThread
        }

        ad.setAdEventListener(object : InterstitialAdEventListener {
            override fun onAdShown() = notifyListeners("interstitialShown", adEvent(interstitialAdUnitId))
            override fun onAdClicked() = notifyListeners("interstitialClicked", adEvent(interstitialAdUnitId))
            override fun onAdImpression(impressionData: ImpressionData?) = notifyListeners("interstitialImpression", adEvent(interstitialAdUnitId))
            override fun onAdDismissed() {
                notifyListeners("interstitialDismissed", adEvent(interstitialAdUnitId))
                destroyInterstitialAd()
            }

            override fun onAdFailedToShow(adError: AdError) {
                notifyListeners("interstitialFailedToShow", errorEvent(interstitialAdUnitId, null, adError.description))
                destroyInterstitialAd()
            }
        })
        ad.show(currentActivity)
        call.resolve()
    }

    @PluginMethod
    fun prepareRewarded(call: PluginCall) = runOnMainThread {
        val adUnitId = requireAdUnitId(call) ?: return@runOnMainThread
        if (!requireInitialized(call)) return@runOnMainThread
        if (rewardedLoadCall != null) {
            call.reject("A rewarded ad load is already in progress", "LOAD_IN_PROGRESS")
            return@runOnMainThread
        }

        destroyRewardedAd()
        rewardedAdUnitId = adUnitId
        rewardedLoadCall = call
        if (rewardedAdLoader == null) rewardedAdLoader = RewardedAdLoader(context)
        val request = AdRequest.Builder(adUnitId).build()
        rewardedAdLoader?.loadAd(request, object : RewardedAdLoadListener {
            override fun onAdLoaded(rewarded: RewardedAd) {
                rewardedAd = rewarded
                notifyListeners("rewardedLoaded", adEvent(rewardedAdUnitId))
                rewardedLoadCall?.resolve()
                rewardedLoadCall = null
            }

            override fun onAdFailedToLoad(error: AdRequestError) {
                val event = errorEvent(rewardedAdUnitId, error.code, error.description)
                notifyListeners("rewardedFailedToLoad", event)
                rewardedLoadCall?.reject(error.description, "REWARDED_LOAD_FAILED")
                rewardedLoadCall = null
            }
        })
    }

    @PluginMethod
    fun isRewardedReady(call: PluginCall) = runOnMainThread {
        call.resolve(JSObject().put("ready", rewardedAd != null))
    }

    @PluginMethod
    fun showRewarded(call: PluginCall) = runOnMainThread {
        val ad = rewardedAd
        val currentActivity = activity
        if (ad == null) {
            call.reject("Rewarded ad is not ready", "AD_NOT_READY")
            return@runOnMainThread
        }
        if (currentActivity == null) {
            call.reject("Android Activity is unavailable", "ACTIVITY_UNAVAILABLE")
            return@runOnMainThread
        }

        ad.setAdEventListener(object : RewardedAdEventListener {
            override fun onAdShown() = notifyListeners("rewardedShown", adEvent(rewardedAdUnitId))
            override fun onAdClicked() = notifyListeners("rewardedClicked", adEvent(rewardedAdUnitId))
            override fun onAdImpression(impressionData: ImpressionData?) = notifyListeners("rewardedImpression", adEvent(rewardedAdUnitId))
            override fun onRewarded(reward: Reward) {
                notifyListeners("rewarded", JSObject().apply {
                    put("adUnitId", rewardedAdUnitId)
                    put("amount", reward.amount)
                    put("type", reward.type)
                })
            }

            override fun onAdDismissed() {
                notifyListeners("rewardedDismissed", adEvent(rewardedAdUnitId))
                destroyRewardedAd()
            }

            override fun onAdFailedToShow(adError: AdError) {
                notifyListeners("rewardedFailedToShow", errorEvent(rewardedAdUnitId, null, adError.description))
                destroyRewardedAd()
            }
        })
        ad.show(currentActivity)
        call.resolve()
    }

    override fun handleOnDestroy() {
        runOnMainThread {
            interstitialLoadCall?.reject("Plugin was destroyed", "PLUGIN_DESTROYED")
            rewardedLoadCall?.reject("Plugin was destroyed", "PLUGIN_DESTROYED")
            initializingCall?.reject("Plugin was destroyed", "PLUGIN_DESTROYED")
            interstitialLoadCall = null
            rewardedLoadCall = null
            initializingCall = null
            interstitialAdLoader?.cancelLoading()
            rewardedAdLoader?.cancelLoading()
            interstitialAdLoader = null
            rewardedAdLoader = null
            destroyInterstitialAd()
            destroyRewardedAd()
        }
    }

    private fun setBoolean(call: PluginCall, setter: (Boolean) -> Unit) = runOnMainThread {
        val value = call.getBoolean("value")
        if (value == null) {
            call.reject("value is required", "INVALID_ARGUMENT")
            return@runOnMainThread
        }
        setter(value)
        call.resolve()
    }

    private fun requireInitialized(call: PluginCall): Boolean {
        if (!initialized) call.reject("Call initialize() before loading ads", "NOT_INITIALIZED")
        return initialized
    }

    private fun requireAdUnitId(call: PluginCall): String? {
        val value = call.getString("adUnitId")?.trim()
        if (value.isNullOrEmpty()) call.reject("adUnitId is required", "INVALID_ARGUMENT")
        return value?.takeIf(String::isNotEmpty)
    }

    private fun versionResult() = JSObject().put("version", YandexAds.libraryVersion)
    private fun adEvent(adUnitId: String) = JSObject().put("adUnitId", adUnitId)
    private fun errorEvent(adUnitId: String, code: Int?, message: String) = JSObject().apply {
        put("adUnitId", adUnitId)
        code?.let { put("code", it) }
        put("message", message)
    }

    private fun destroyInterstitialAd() {
        interstitialAd?.setAdEventListener(null)
        interstitialAd = null
    }

    private fun destroyRewardedAd() {
        rewardedAd?.setAdEventListener(null)
        rewardedAd = null
    }

    private fun runOnMainThread(block: () -> Unit) {
        activity?.runOnUiThread(block) ?: bridge.executeOnMainThread(block)
    }
}
